File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Compare {
public:
    bool operator()(ListNode* l1, ListNode* l2) {
        return l1->val >= l2->val;
    }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, Compare> pq;
        for (auto list: lists) {
            if (list) {
                pq.push(list);
            }
        }
        ListNode* head = nullptr;
        ListNode* curr;
        ListNode* prev;
        while (!pq.empty()) {
            curr = pq.top();
            pq.pop();
            if (head == nullptr) {
                head = curr;
            } else {
                prev->next = curr;
            }
            if (curr->next) {
                pq.push(curr->next);
            }
            prev = curr;
        }
        return head;
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) {
            return null;
        }
        PriorityQueue<ListNode> pq = new PriorityQueue<>(lists.length, new ListNodeComparator());
        for (int i = 0; i < lists.length; i++) {
            if (lists[i] != null) {
                pq.add(lists[i]);
            }
        }
        ListNode head = null;
        ListNode prev = null;
        ListNode curr = null;
        while (!pq.isEmpty()) {
            curr = pq.poll();
            if (head == null) {
                head = curr;
            }
            if (prev != null) {
                prev.next = curr;
            }
            if (curr.next != null) {
                pq.add(curr.next);
            }
            prev = curr;
        }
        return head;
    }
}

class ListNodeComparator implements Comparator<ListNode> {
    @Override
    public int compare(ListNode l1, ListNode l2) {
        return l1.val - l2.val;
    }
}